### DISCUSSION TITLE
That "Beginners Guide" wiki page...

### DISCUSSION THREAD

**Discussion by OtherJohnGray** [zbm-dev/zfsbootmenu#343] [General]
Hi everyone, I'm John, the guy who has been working on the "Complete Beginner's Guide" page in the wiki. I thought that I should say hi and introduce myself / explain why it's there and where it's up to, since it's been taking forever to finish and Andrew has had to put a disclaimer at the top of it...

I started writing it in response to this discussion with u/E39M5S62 on the Reddit ZBM 2.0 announcement post:
https://www.reddit.com/r/zfs/comments/vmqw63/comment/iewc6tf/
(I'm not sure if E39M5S62 is one of you guys?)

As per that discussion, I'm delighted to have access to a boot manager that can manage/recover my ZFS boot environments, and I thought it might help if I put together an example of the type of detailed step-by-step documentation that I thought could help end-users like myself adopt ZBM with Root-on-ZFS more easily, either as a suggestion for other documentation, or if I managed to get it right, as an additional documentation asset. I have been building it in the wiki rather than making a markdown pull request because all the other end-user guides seem to be there too.

It took me a long time to write because I have been learning as I go, and there are some parts I did not understand at first and that I need to rewrite, in particular the part about having to double-enter the pool decryption password and putting the decryption key in the initramfs instead - I have since got that working, an I think that's a much better way to go that what I first wrote. I think it could also use an appendix with how to optionally build ZBM with dropbear for ssh access rather than just using the pre-build EFI image, and also how to set up a script to mount swap on a degraded mdraid array if one of the raid1 partitions is unavailable at boot.

I still need to give it an end-to-end test by using it to install ZBM onto a spare machine, to make sure that I have everything in order and that following the instructions actually results in a working system - hopefully I should be able to find time for that over the next few weeks. 

Does what I have written seem to be roughly on-track, and like it might be of some use?

**Comment by ahesford** [comment]
We appreciate the contribution. It is off to a good start and I'm sure with a few refinements it will be a resource for those interested in Debian. The plan is to remove the notice of unofficiality once one of us has actually been through the full guide and are comfortable recommending to people looking for help. In the meantime, neither of us has had the time to review everything.

I started making some edits because somebody on IRC was having difficulty following your guide. I wanted to clean up some formatting and clarify a few points. Also, the guide originally overlooked the mounting of filesystems you created; because the root filesystem is `canmount=off` (which I changed to `canmount=noauto`; anything you intend to call `zfs mount` on should be `noauto` instead of `off`), none of the filesystems to which the install should be made are actually mounted.

Regarding formatting, I swapped some explicit URLs for more descriptive hyperlinks and started making liberal use of code ticks to designate "machine" aspects when talking about them in text. For example, instead of just referring to "/dev/sda" (or, worse, "/dev/sd\*"), surround in backticks to get `/dev/sda` and `/dev/sd*`. These callouts make it obvious these are machine parameters that should be understood from system output or provided as system input. It's also useful to make sure monospace fonts are used, avoiding the potential for confusion.

I don't mind the intent of the tables explaining command arguments, but I'm not sure if that's the best format for presenting the information; especially when I started back-quoting the parameters in the left column, the tables started to wrap weird.

My edits currently run through the end of Section 2.

**Comment by OtherJohnGray** [comment]
Sorry about the canmount, that's the type of thing I was hoping to pick up with a full run-through. I was going back and revising some sections as I found mistakes in what I had done, so I'm not sure that I captured everything.

Shall I finish the rest of the code ticks and descriptive hyperlinks? 

I'm not so sure about the argument tables anymore either. I think there is value in explaining what various cryptic commands do, but they are kind of starting to dominate the document. Do you think they might be better hidden in collapsed detail tags, perhaps with a summary of "(command explanation)", and maybe just make them paragraphs with **bold descriptions: ** or something like that? Otherwise I'm not sure if it would be best to keep them as is or remove them all together.

**Comment by ahesford** [comment]
Yeah, you can finish the reformat and do whatever you think is appropriate with respect to the tables.

**Reply by OtherJohnGray** [reply]
I've put in the remaining back-ticks, and I have collapsed the command explanation tables into HTML details blocks with "(click here for an explanation of the commands above)" as a summary. 

Next up, I guess I should probably in-line the instructions for putting the ZFS decryption key into the initramfs, and then give it an end-to-end walk through on a test machine to pick up any remaining errors. I should be able to find time for that this week.

**Comment by OtherJohnGray** [comment]
Status update: 
I've in-lined the instructions for adding the ZFS encryption key to initramfs, and adjusted the rest of the text accordingly.

I started an end-to-end test, but quickly ran into a problem with `sgdisk` thinking a partition on the live USB had been modified while in use, and then not loading partition table changes on the other drives. I'm currently trying to find time to isolate the cause by bisecting the steps between the first and second calls to `sgdisk` to see what exactly is initiating the problem and maybe what the cause is:

    root@debian:/home/user# sgdisk -n2:0:-65544M -t2:BF00 -c 2:"Root-0-SATA" $disk1
    sgdisk -n2:0:-65544M -t2:BF00 -c 2:"Root-1-NVMe" $disk2
    Setting name!
    partNum is 1
    Warning: The kernel is still using the old partition table.
    The new table will be used at the next reboot or after you
    run partprobe(8) or kpartx(8)
    The operation has completed successfully.
    Setting name!
    partNum is 1
    Warning: The kernel is still using the old partition table.
    The new table will be used at the next reboot or after you
    run partprobe(8) or kpartx(8)
    The operation has completed successfully.
    root@debian:/home/user# ls -l /dev/disk/by-partlabel/
    total 0
    lrwxrwxrwx 1 root root 10 Nov  6 07:55 EFI-0-SATA -> ../../sda1
    lrwxrwxrwx 1 root root 15 Nov  6 07:55 EFI-1-NVMe -> ../../nvme0n1p1
    root@debian:/home/user# kpartx
    bash: kpartx: command not found
    root@debian:/home/user# partprobe
    Error: Partition(s) 1 on /dev/sdb have been written, but we have been unable to inform the kernel of the change, probably because it/they are in use.  As a result, the old partition(s) will remain in use.  You should reboot now before making further changes.
    root@debian:/home/user# lsblk
    NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
    loop0         7:0    0   2.3G  1 loop /usr/lib/live/mount/rootfs/filesystem.squashfs
    sda           8:0    0 931.5G  0 disk 
    ├─sda1        8:1    0   512M  0 part /tmp/efi0
    └─sda2        8:2    0   867G  0 part 
    sdb           8:16   1  14.9G  0 disk 
    ├─sdb1        8:17   1   2.7G  0 part /usr/lib/live/mount/medium
    └─sdb2        8:18   1   2.6M  0 part 
    nvme0n1     259:0    0 931.5G  0 disk 
    ├─nvme0n1p2 259:1    0   867G  0 part 
    └─nvme0n1p1 259:2    0   512M  0 part /tmp/efi1

**Reply by OtherJohnGray** [reply]
Clarification: The issue seems to occur between the first two calls to sgdisk to create the EFI partitions, which succeed  (not shown in above exampe), and the second two calls to sgdisk to create the root partitions (shown above), which fail to notify the kernel.

**Comment by zdykstra** [comment]
Your machine is missing `kpartx`. Being able to successfully run that will
help.

**Reply by OtherJohnGray** [reply]
I found the issue - weirdly if the EFI partitions are mounted when I run the second pair of `sgdisk` commands, then I get the error, if I unmount the partitions first, then I don't get the error. I guess it's probably to do with having something connected to a mountpoint directory that was defined on the filesystem that was already on the flash drive (sdb), which is causing sgdisk to take a look at the partition table on sdb and decide it doesn't like it, perhaps because it's ISO9660 hybrid? 

Anyway, I guess I should just update the document to have those filesystems unmounted before proceeding with the second set of sgdisks, and start over with the end-to-end test...

**Comment by OtherJohnGray** [comment]
Hi,

After many end-to-end tests and corrections, I am finally confident that following the steps in the document will result in a working system, at least On My Machine(tm). 

I have also been using another machine, that I built using an early version of the document, as a daily driver for a couple of months so I'm also reasonably confident that the system you end up with is suitable for general use.

I think I'll leave it there for now. It doesn't cover building an EFI image with remote access, but I have linked to the new "Building in Containers" page which does, so hopefully that will meet the needs of anyone who requires that. 

I took the liberty of removing "still a work in progress" from the notice at the start, as that didn't seem correct any more - hope that was right?

**Reply by zdykstra** [reply]
Thank you for all of the time and energy you've put into that guide. It's greatly appreciated! I'll read through it again to see if there are any tweaks that should be made for the upcoming 2.1.0 release.

Linking to the documentation on containerized builds for remote access to ZBM is perfect; there's no need to repeat what's already documented on other pages.
